src/geometry/planar.md

```diff
@@ -0,0 +1,347 @@
+                return pl.half() < pr.half();
+            return pl.cross(pr) > 0;
+        };
+        std::sort(adj[i].begin(), adj[i].end(), compare);
+    }
+    std::vector<std::vector<size_t>> faces;
+    for (size_t i = 0; i < n; i++) {
+        for (size_t edge_id = 0; edge_id < adj[i].size(); edge_id++) {
+            if (used[i][edge_id]) {
+                continue;
+            }
+            std::vector<size_t> face;
+            size_t v = i;
+            size_t e = edge_id;
+            while (!used[v][e]) {
+                used[v][e] = true;
+                face.push_back(v);
+                size_t u = adj[v][e];
+                size_t e1 = std::lower_bound(adj[u].begin(), adj[u].end(), v, [&](size_t l, size_t r) {
+                    Point pl = vertices[l] - vertices[u];
+                    Point pr = vertices[r] - vertices[u];
+                    if (pl.half() != pr.half())
+                        return pl.half() < pr.half();
+                    return pl.cross(pr) > 0;
+                }) - adj[u].begin() + 1;
+                if (e1 == adj[u].size()) {
+                    e1 = 0;
+                }
+                v = u;
+                e = e1;
+            }
+            std::reverse(face.begin(), face.end());
+            int sign = 0;
+            for (size_t j = 0; j < face.size(); j++) {
+                size_t j1 = (j + 1) % face.size();
+                size_t j2 = (j + 2) % face.size();
+                int64_t val = vertices[face[j]].cross(vertices[face[j1]], vertices[face[j2]]);
+                if (val > 0) {
+                    sign = 1;
+                    break;
+                } else if (val < 0) {
+                    sign = -1;
+                    break;
+                }
+            }
+            if (sign <= 0) {
+                faces.insert(faces.begin(), face);
+            } else {
+                faces.emplace_back(face);
+            }
+        }
+    }
+    return faces;
+}
+```
+
+## Building planar graph from line segments
+
+Sometimes you are not given a graph explicitly, but rather as a set of line segments on a plane, and the actual graph is formed by intersecting those segments, as shown in the picture below. In this case you have to build the graph manually. The easiest way to do so is as follows. Fix a segment and intersect it with all other segments. Then sort all intersection points together with the two endpoints of the segment lexicographically and add them to the graph as vertices. Also link each two adjacent vertices in lexicographical order by an edge. After doing this procedure for all edges we will obtain the graph. Of course, we should ensure that two equal intersection points will always correspond to the same vertex. The easiest way to do this is to store the points in a map by their coordinates, regarding points whose coordinates differ by a small number (say, less than $10^{-9}$) as equal. This algorithm works in $O(n^2 \log n)$.
```

Comment:
I feel like there should be some more details about how to use map for this... Do we just round both coordinates before inserting a point into the map to ensure that "equal" points end up in the same bucket? Are there better alternatives? Does it work adequately in practice?

Reply:
In the implementation I overload the comparators for the `Point` class to tolerate at most `eps` difference in coordinates. This way I passed one of the linked problems which required to build the graph from segments so I guess this works adequately. Original e-maxx code follows similar approach, but I don't know whether it was tested anywhere.

Reply:
I'm a bit wary about using non-transitive comparators... Wouldn't it be better to round the points to a certain precision to ensure associativity? At least this way we would somewhat ensure the strict weak ordering.

Reply:
Yes, you are right. I changed the code, now the coordinates are multiplied by 1e9 and rounded to the nearest integer.

Reply:
Thanks! Could you also update the text description to mirror it?